compiler/AST/AstToText.cpp

```diff
@@ -896,7 +898,139 @@ static bool looksLikeInfixOperator(const char *fnName)
+           strcmp(op, ">") == 0 || strcmp(op, ">=") == 0)
+    return 5;
+  else if (strcmp(op, "==") == 0 || strcmp(op, "!=") == 0)
+    return 4;
+  else if (strcmp(op, "&&") == 0)
+    return 3;
+  else if (strcmp(op, "||") == 0)
+    return 2;
+  // by and align are precedence 1 too, but don't come through this path.
+  else if (strcmp(op, "#") == 0)
+    return 1;
+  printf("unknown op %s", op);
+  return 0;
+}
+
+/*
+ * needParens(outer, inner, ...) is called to evaluate whether we need
+ * to add parens around the inner op.  We're here when
+ * appendExpr(outer) is calling appendExpr(expr->get(1)) or
+ * expr->get(2), which has become appendExpr(inner).
+ *
+ * Given an AST node outer, with one child inner, this function
+ * returns true if we need to emit parens around the child expression.
+ * That is, if emitting the expression without parenthesis would
+ * change the semantics from what the AST represents.
+ *
+ * If the inner (child) operator has higher precedence than the outer
+ * (parent), then we don't need parens, as emitting
+ * "a outer_op b inner_op c" is equivalent to
+ * "a outer_op (b inner_op c)".
+ *
+ * If the child operator has equal precedence to the outer, then we
+ * generally don't need parenthesis, except for a few exceptions.  If
+ * the desired expression is a-(b-c) or a-(b+c), then the parentheses
+ * are needed.  Of course they're not needed for a+(b-c) or a+(b+c).
+ * If the desired expression is a/(b/c) or a/(b*c) or a%(b/c), they're
+ * needed, but not for a*(b*c) or a*(b/c).
+ * (TODO: a*(b/c) might need the parens for overflow reasons.)
+ *
+ * Unary - (and +) have higher precedence than binary - (and +), so we
+ * need the unary flag to know which case we're in.
+ *
+ * The postfix flag is needed because postfix ! has higher precedence
+ * than prefix !.
+ *
+ * The innerIsRHS flag is needed because we need parens to express
+ * a-(b-c) where inner- is the RHS of the outer-.  But we can emit
+ * a-b-c instead of (a-b)-c -- when inner- is the LHS of outer-.
+ */
+static bool needParens(const char *outer, const char *inner,
+                       bool outerUnary, bool outerPostfix, bool innerIsRHS) {
+  bool ret = false;
+  int outerprec, innerprec;
+
+  if (!outer)
+    return false;
+
+  outerprec = opToPrecedence(outer, outerUnary, outerPostfix);
+  innerprec = opToPrecedence(inner, false, false); // inner is a binary op
+
+  if (outerprec > innerprec)
+    ret = true;
+
+  // If inner and outer have the same precedence, and inner is the
+  // rhs, it needs parens if a op1 (b op2 c) isn't equivalent to
+  // a op1 b op2 c.  (Note op1 and op2 may be the same op, a - (b - c))
+  if (innerIsRHS &&
+      (strcmp(outer, "-") == 0 ||
+       strcmp(outer, "/") == 0 || strcmp(outer, "%") == 0 ||
+       strcmp(outer, "<<") == 0 || strcmp(outer, ">>") == 0 ||
+       // (a==b)==true vs. a==(b==true)
+       strcmp(outer, "==") == 0 || strcmp(outer, "!=") == 0)
```

Comment:
These two can be comparisons to astrSeq and astrSne respectively instead of performing a strcmp

Reply:
Partial progress is the reason.  We encountered some cases that it was really beneficial to convert and haven't gotten to the rest just yet.  Hopefully they'll have astrs eventually